vespa-http-client/src/test/java/com/yahoo/vespa/http/client/core/communication/IOThreadTest.java

```diff
@@ -124,4 +142,59 @@ public void testQueueTimeOutNoNoConnectionToServer() throws Exception {
             assert (latch.await(120, TimeUnit.SECONDS));
         }
     }
+
+    @Test
+    public void requireThatEndpointProtocolExceptionsArePropagated()
+            throws IOException, ServerResponseException, InterruptedException, TimeoutException, ExecutionException {
+        when(apacheGatewayConnection.connect()).thenReturn(true);
+        int errorCode = 403;
+        String errorMessage = "Not authorized";
+        doThrow(new ServerResponseException(errorCode, errorMessage)).when(apacheGatewayConnection).handshake();
+        Future<FeedEndpointException> futureException = endpointErrorCapturer(endpointResultQueue);
+
+        try (IOThread ioThread = new IOThread(
+                endpointResultQueue, apacheGatewayConnection, 0, 0, 10, 10L, documentQueue, 0)) {
+            ioThread.post(doc1);
+            FeedEndpointException reportedException = futureException.get(120, TimeUnit.SECONDS);
+            assertThat(reportedException, instanceOf(FeedProtocolException.class));
+            FeedProtocolException actualException = (FeedProtocolException) reportedException;
+            assertThat(actualException.getHttpStatusCode(), equalTo(errorCode));
+            assertThat(actualException.getHttpResponseMessage(), equalTo(errorMessage));
+            assertThat(actualException.getEndpoint(), equalTo(ENDPOINT));
+            assertThat(actualException.getMessage(), equalTo("Endpoint 'myhost:4080' returned an error on handshake: 403 - Not authorized"));
+        }
+    }
+
+    @Test
+    public void requireThatEndpointConnectExceptionsArePropagated()
+        throws IOException, ServerResponseException, InterruptedException, TimeoutException, ExecutionException {
+        when(apacheGatewayConnection.connect()).thenReturn(true);
+        String errorMessage = "generic error message";
+        IOException cause = new IOException(errorMessage);
+        doThrow(cause).when(apacheGatewayConnection).handshake();
+        Future<FeedEndpointException> futureException = endpointErrorCapturer(endpointResultQueue);
+
+        try (IOThread ioThread = new IOThread(
+                endpointResultQueue, apacheGatewayConnection, 0, 0, 10, 10L, documentQueue, 0)) {
+            ioThread.post(doc1);
+            FeedEndpointException reportedException = futureException.get(120, TimeUnit.SECONDS);
+            assertThat(reportedException, instanceOf(FeedConnectException.class));
```

Comment:
Related to the above comment, should this instead be a `FeedHandshakeException`?